Auth feature

kajarenc opened the pull request:
## Describe your changes

This feature allows to configure Auth in the streamlit app with arbitrary OIDC providers (like Google, Microsoft Entra ID, etc.). 

The demo app repo: https://github.com/kajarenc/stauthlib

## GitHub Issue Link (if applicable)

Closes #8518 

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python) ADDED
- E2E Tests ADDED
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

kajarenc committed:
draft experiments for auth workflows

kajarenc committed:
add login via st.user.login()

kajarenc committed:
Merge branch 'develop' into auth-experiments

kajarenc committed:
Add ability to send message to HOST frame from st.user.login()

kajarenc committed:
Add logout functionality

kajarenc committed:
Add logout functionality

kajarenc committed:
Add logout functionality,
Rename cookie name to `_streamlit_uzer`

kajarenc committed:
use redirect_uri from secrets

kajarenc committed:
revert hostframe changes

kajarenc committed:
Implementation that uses Authlib

kajarenc committed:
Add OIDC implementation

kajarenc committed:
empty commit to trigger CI

kajarenc committed:
Remove unnecessary prints

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

# Conflicts:
#	frontend/app/src/App.tsx

kajarenc committed:
Merge with latest develop (MPA)

kajarenc committed:
Merge branch 'refs/heads/develop' into auth-authlib-experiments

kajarenc committed:
rename `/login` endpoint to `/oauth2callback`

kajarenc committed:
either send message to Host frame or do redirect based on do we in child iframe

kajarenc committed:
run streamlit app if auth don't configured

kajarenc committed:
remove old Tornado built-in oauth handlers

kajarenc committed:
Merge branch 'refs/heads/develop' into auth-authlib-experiments

kajarenc committed:
Merge branch 'refs/heads/develop' into auth-authlib-experiments

# Conflicts:
#	lib/streamlit/web/server/browser_websocket_handler.py

kajarenc committed:
Merge branch 'refs/heads/develop' into auth-authlib-experiments

kajarenc committed:
include all information from user id token to st.experimental_user

kajarenc committed:
put only user email instead of whole user id token in for to st.experimental_user

kajarenc committed:
Merge branch 'refs/heads/develop' into auth-authlib-experiments

# Conflicts:
#	frontend/app/src/App.tsx

kajarenc committed:
small fixes, todos added

kajarenc committed:
Merge branch 'refs/heads/develop' into auth-authlib-experiments

# Conflicts:
#	lib/streamlit/user_info.py

kajarenc committed:
add all available userinfo from identity token to st.experimental_user

kmcgrady commented on lib/streamlit/web/server/authlib_tornado_integration.py:
```diff
@@ -0,0 +1,32 @@
+            v = oauth.config.get(name, {}).get(k, None)
+            if v is not None:
+                rv[k] = v
+        return rv
```
What do these things do? I assume it's authlib specific. Will benefit from some comments.

kajarenc replied:
Yep, I refactored it a little bit to have better names and docsting.
In a nutshell, we need to override how authlib reads configs to be able to specify multiple auth providers based on secrets.toml definition

kmcgrady commented on lib/streamlit/web/server/oauth_authlib_routes.py:
```diff
@@ -0,0 +1,122 @@
+oauth = TornadoOAuth(config, cache=my_cache)
+
+if secrets_singleton.load_if_toml_exists():
+    auth_section = secrets_singleton.get("auth", None)
```
With recent changes to secrets, should we assume the auth might update over time? Perhaps for key rotating or something?

kmcgrady reviewed (commented):
Overall this is looking good. Gave a couple comments to understand deeper.

kajarenc committed:
add explanatory comment

kajarenc committed:
cast secrets AttrDict nested sections to normal dict

kajarenc committed:
Merge branch 'refs/heads/develop' into auth-authlib-experiments

kajarenc committed:
load secrets lazily
validate on user.login()

kajarenc committed:
refactor provider registration logic

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Added error state handling when user doesn't provide consent

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Add is_logged_in method for st.experimental_user

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Added CSRF validations for auth user info cookie

kajarenc committed:
Added short living token for encrypting login start endpoint. this is needed to prevent starting auth flow outside of streamlit app

kajarenc committed:
Finalize naming

kajarenc committed:
Send message for auth request only to the same origin parent iframe

kajarenc committed:
Add authlib to dependency list to run tests

kajarenc committed:
Pass origin also in case of error

kajarenc committed:
fix typing issues

kajarenc committed:
fix typing issues

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Change default prompt strategy to "select_account"

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
use get_secure cookie if old version of Tornado is used

kajarenc committed:
update min dependency list, add Authlib

kajarenc committed:
update min dependency list, add Authlib

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Add e2e test for auth flow, skip at the moment before we understand how todo secret management for test

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
add unit tests

kajarenc committed:
clear user info on ScriptRunContext reset

kajarenc committed:
fix missing user data error

kajarenc committed:
fix CodeQL warning

kajarenc committed:
add more unit tests

kajarenc committed:
add more unit tests

kajarenc committed:
Add more unit tests (auth login endpoint test)

kajarenc committed:
remove redundant proto import

mauro-mascia commented on lib/streamlit/web/server/oauth_authlib_routes.py:
```diff
@@ -0,0 +1,163 @@
+        provider_client_kwargs["scope"] = "openid email profile"
+    if "prompt" not in provider_client_kwargs:
+        provider_client_kwargs["prompt"] = "select_account"
+
```
for me, using Microsoft AD, I need to also have this row:
```
provider_client_kwargs["code_challenge_method"] = "S256"
```
to enable the generation of the PKCE code.
Can this be added as parameter in the next relase?

kajarenc replied:
Hey @mauro-mascia, thank you for your feedback!

We already set "code_challenge_method" to "S256": if it is supported by OIDC provider, we look to `code_challenge_methods_supported` value for `.well-known` configuration. 
So most probably it is already work. 

https://github.com/streamlit/streamlit/pull/8786/files#diff-e7111ded76c5b321c2808266062e60890d8bca6d1b202533ef2e6ec62befd5a8R33-R37

Also you can always set client_kwargs explicitly in secrets.toml auth section of concrete provider:
```toml
[auth]
redirect_uri = "http://localhost:8501/oauth2callback"
cookie_secret = "your_cookie_secret_here"

[auth.auth0]
client_id = "AAA"
client_secret = "BBB"
server_metadata_url = "https://cccc.us.auth0.com/.well-known/openid-configuration"
client_kwargs = { 'prompt' = 'consent' }
```

In the next weeks we will work to make it more explicit!

mauro-mascia replied:
thanks @kajarenc using ```client_kwargs = { 'code_challenge_method' = 'S256' }``` was the way to go

mauro-mascia replied:
My last issue is that when I return from the callback I correctly got the `code=1.ASAAGgwVWwGhb...` (I'm on Microsoft AD) but it does not store any cookie. I think that `AuthCallbackHandler` should be called but it does not go here. Any idea?

kajarenc replied:
Is there any chance you are behind the proxy? 
You should configure `redirect_uri` in the `secrets.toml`

Something like: 
```toml
[auth]
redirect_uri = "http://localhost:8501/oauth2callback"
cookie_secret = "your_cookie_secret_here"

[auth.microsoft]
client_id = "blablabla"
```
and the `redirect_uri` should be the same as that configured in your Microsoft AD app. 

Also, does Microsoft AD support OIDC protocol?
Because this feature is intended to support only OIDC, I tested it with `Microsoft Entra ID`, and it works with it.

mauro-mascia replied:
Thanks @kajarenc this question was related with another comment https://github.com/streamlit/streamlit/pull/8786#discussion_r1838506244
and the fact that the redirect has been configured to http://localhost:8501/callback. For that reason it wasn't working. Maybe we need to specify that this is the required return URL somewhere if not already specified (I didn't see it until watching the code). Anyway it's ok for me to set it to /oauth2callback thanks.

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
mock network interaction

kajarenc committed:
add more test

mauro-mascia commented on lib/streamlit/web/server/oidc_mixin.py:
```diff
@@ -0,0 +1,102 @@
+        claims_options = kwargs.pop("claims_options", None)
+        state_data = self.framework.get_state_data(session, params.get("state"))
+        self.framework.clear_state_data(session, params.get("state"))
+        params = self._format_state_params(state_data, params)
```
@kajarenc  After this line I need to add

```
        # Get redirect_uri from state_data or params
        if 'headers' not in kwargs:
            kwargs['headers'] = {}
        redirect_uri = state_data.get('redirect_uri') or params.get('redirect_uri')
        kwargs['headers']['Origin'] = redirect_uri
```

to avoid `authlib.integrations.base_client.errors.OAuthError: invalid_request: AADSTS9002327: Tokens issued for the 'Single-Page Application' client-type may only be redeemed via cross-origin requests.`

kajarenc replied:
It would be unsafe, I think specifying correct same `redirect_uri` both `secrets.toml` and Microsoft app should solve the problem.

kajarenc committed:
add more test

mauro-mascia commented on lib/streamlit/web/server/server.py:
```diff
@@ -95,6 +105,10 @@
     r"(?:script-health-check|_stcore/script-health-check)"
 )
 
+OAUTH2_CALLBACK_ENDPOINT: Final = "/oauth2callback"
```
@kajarenc should be cool to be configurable with toml

kajarenc replied:
At this stage, we don't plan to make it configurable, since this required during Tornado app initialization.

kajarenc committed:
add logout tests

kajarenc committed:
Add tests for AuthCallbackHandler handler

kajarenc committed:
mock more granular: `set_signed_cookie` instead of `set_auth_cookie`

kajarenc committed:
use fallback if `set_signed_cookie` not available (for Tornado versions <6.3.0)

kajarenc committed:
Add more tests

kajarenc committed:
Add explanatory comment

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
add tests for oidc_mixin.py

kajarenc committed:
sort imports

kajarenc committed:
add e2e tests

kajarenc committed:
change port to 8555 to run e2e test

kajarenc committed:
change app port to 7999

kajarenc committed:
change app port to 8501

kajarenc committed:
add better fixtures

kajarenc committed:
get available port for oidc_mock_server implementation

kajarenc committed:
Add js unit tests

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
fix js imports in tests

kajarenc committed:
empty commit to trigger rerun

kajarenc committed:
add util code from e2e_playwright/shared to changed file detection

kajarenc committed:
try another pattern

kajarenc committed:
add also test secrets

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
fix docstring

kajarenc committed:
add docstrings

kajarenc committed:
add docstrings

kajarenc committed:
add docstrings

kajarenc committed:
use tempfile instead of hardcoded secrets

kajarenc committed:
bunch of small refinements

kmcgrady commented on frontend/app/src/App.tsx:
```diff
@@ -717,6 +718,16 @@ export class App extends PureComponent<Props, State> {
+            this.hostCommunicationMgr.sendMessageToSameOriginHost({
+              type: "REDIRECT_TO_URL",
+              url: authRedirect.url,
+            })
```
Do we want to consider the case of embeds (not Streamlit Cloud), is it worth redirecting anyways in this case?)

kajarenc replied:
Sorry, forgot to respond here :) 

yes, in this case, instead of standard `sendMessageToHost`, we use `sendMessageToSameOriginHost` to ensure that just by embedding a popular streamlit app into your iframe, you will not be able to intercept an authentication request. 

And we send the message to the parent host, so other Streamlit hosting providers would be able to implement their changes to make native streamlit authentication work on their platform.

kmcgrady commented on lib/streamlit/user_info.py:
```diff
@@ -14,16 +14,118 @@
+            "To use Auth you need to install the 'Authlib' package."
+        ) from None
+
+    claim_options = {"exp": {"essential": True}, "provider": {"essential": True}}
```
Might want to add a comment that explains these claim options.

kajarenc replied:
Done

kmcgrady commented on lib/streamlit/user_info.py:
```diff
@@ -14,16 +14,118 @@
+        raise StreamlitAPIException(
+            f"Auth credentials for '{provider}' are missing the following keys: "
+            f"{missing_keys}. Please check your configuration."
+        )
```
We are getting into the case of creating custom error messages, should we either:
1. Define custom exception classes for these.
2. Define a single exception to encapsulate Auth config issues here

kajarenc replied:
I defined a specific AuthError, and then raise that exception when previously raised StreamlitAPIException

kmcgrady commented on lib/streamlit/user_info.py:
```diff
@@ -62,6 +164,39 @@ class UserInfoProxy(Mapping[str, Union[str, None]]):
+
+            if runtime.exists():
+                instance = runtime.get_instance()
+                session_info = instance._session_mgr.get_session_info(session_id)
```
We should probably define methods on the runtime rather than accessing the "private" variable here something like `set_user_info(session_id, user_info)` and/or `clear_user_info_for_session(session_id)`

kajarenc replied:
done

kmcgrady commented on lib/streamlit/user_info.py:
```diff
@@ -62,6 +164,39 @@ class UserInfoProxy(Mapping[str, Union[str, None]]):
+        user_email = _get_user_info().get("email")
+        if user_email is not None and user_email != "test@example.com":
+            return True
+        return False
```
Nit: Can simplify `return user_email is not None and user_email != "test@example.com"`

kajarenc replied:
done

kmcgrady commented on lib/streamlit/web/server/browser_websocket_handler.py:
```diff
@@ -50,13 +57,61 @@ def initialize(self, runtime: Runtime) -> None:
+        Tornado < 6.3.0.
+        """
+        try:
+            return super().get_signed_cookie(name, value, max_age_days, min_version)
```
What's the error scenario this is designed to handle? Would just want a comment on the problem the try except is trying to handle.

kmcgrady commented on lib/streamlit/web/server/browser_websocket_handler.py:
```diff
@@ -50,13 +57,61 @@ def initialize(self, runtime: Runtime) -> None:
+
+        if not decoded_token or not decoded_expected_token:
+            return False
+        return hmac.compare_digest(decoded_token, decoded_expected_token)
```
Nit: This seems like this could just be.

```python
return decoded_token and decoded_expected_token and hmac.compare_digest(decoded_token, decoded_expected_token)

kmcgrady commented on lib/streamlit/web/server/oauth_authlib_routes.py:
```diff
@@ -0,0 +1,201 @@
+        """Redirect to the OAuth provider login page."""
+        provider = self._parse_provider_token()
+        if provider is None:
+            self.redirect("/")
```
Does this work properly for `server. baseUrlPath` is set? Might want to just verify.

kmcgrady commented on lib/streamlit/web/server/oauth_authlib_routes.py:
```diff
@@ -0,0 +1,201 @@
+    def _get_origin_from_secrets(self) -> str | None:
+        redirect_uri = None
+        if secrets_singleton.load_if_toml_exists():
+            auth_section = secrets_singleton.get("auth")
```
I'm seeing repetition from above (load_if_toml_exists, etc), would it make sense to have a helper function get all the auth as a dictionary or something and simplify?

Can be as easy as:

```python
if auth_section := get_auth_section():
  redirect_uri = auth_section.get("redirect_uri", None)
```

Can even do better with an argument of the field you wish to get.

```python
if not (redirect_uri := get_auth_value("redirect_uri")):
    return None
    
redirect_uri_parsed = urlparse(redirect_uri)
```

This could be on the secrets_singleton as opposed to a standalone function, but either seems fine.

kmcgrady commented on lib/streamlit/web/server/oidc_mixin.py:
```diff
@@ -0,0 +1,105 @@
+        state = kwargs.pop("state", None)
+        if state:
+            assert self.framework.cache is not None
+            session = None
```
I think we should have a comment. I believe the session is not used by Authlib because it should be retrieving it from the cache?

kmcgrady commented on lib/tests/streamlit/user_info_test.py:
```diff
@@ -118,3 +142,81 @@ def test_st_user_reads_from_context_(self):
+        assert parsed_payload["provider"] == provider
+
+    def test_user_login_with_invalid_provider(self):
+        """Test that st.experimental_user.login sends correct proto message."""
```
Duplicate comment, should change to reflect.

kajarenc replied:
done

kmcgrady commented on lib/tests/streamlit/user_info_test.py:
```diff
@@ -118,3 +142,81 @@ def test_st_user_reads_from_context_(self):
+
+        assert "Auth credentials are missing *'invalid_provider'*" in str(ex.exception)
+
+    def test_user_login_redirect_uri_missing(self):
```
I'd recommend having a description for each test of what to expect. Something like "Tests that an error is raised if the redirect uri is missing"

I didn't check all of the tests but just make sure the relevant info is there.

kajarenc replied:
done

kmcgrady commented on lib/tests/streamlit/user_info_test.py:
```diff
@@ -118,3 +142,81 @@ def test_st_user_reads_from_context_(self):
+            )
+
+    def test_user_logout(self):
+        """Test that st.experimental_user.login sends correct proto message."""
```
should be logout

kajarenc replied:
fixed

kmcgrady commented on lib/tests/streamlit/web/server/oauth_authlib_routes_test.py:
```diff
@@ -0,0 +1,259 @@
+    def test_auth_callback_failure_missing_provider_in_cache(
+        self, mock_set_auth_cookie
+    ):
+        """Test auth callback success."""
```
failure?

kajarenc replied:
fixed

kmcgrady commented on e2e_playwright/auth_test.py:
```diff
@@ -0,0 +1,102 @@
+    wait_for_app_run(app)
+
+    text = get_markdown(app, "John Doe")
+    expect(text).to_be_visible()
```
Can we make a test_login_failure scenario?

kajarenc replied:
Done!

kmcgrady reviewed (commented):
This PR is really strong. Most of my comments help clean up and simplify.

kajarenc committed:
fixes after review

kajarenc committed:
move user info removal on logout to runtime / app session methods

kajarenc committed:
respect base url in more places

kajarenc committed:
fix failing tests

kajarenc committed:
use custom error class for Auth flow errors

kajarenc committed:
add failure e2e test

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
move auth utils to separate file

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
move more functions to auth_util

kajarenc committed:
make mypy happy

kajarenc committed:
add more auth_util tests

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
add docstrings

kajarenc committed:
add gather_metrics usage

kajarenc committed:
check is user logged in not based on email

kajarenc committed:
fix failing test

kajarenc committed:
fix failing test

kajarenc committed:
make st.experimental_user flat object

kajarenc committed:
rename is_logged_in to property is_authenticated

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

sfc-gh-kmcgrady commented on lib/setup.py:
```diff
@@ -49,6 +49,7 @@
     "rich>=10.14.0, <14",
     "tenacity>=8.1.0, <10",
     "toml>=0.10.1, <2",
+    "Authlib>=1.3.1",
```
nit: Can we put this in alphabetical order?

kajarenc replied:
done

sfc-gh-kmcgrady commented on lib/streamlit/runtime/runtime.py:
```diff
@@ -335,7 +346,7 @@ def is_active_session(self, session_id: str) -> bool:
         self,
         client: SessionClient,
-        user_info: dict[str, str | None],
+        user_info: dict[str, str | bool | None],
```
Can you help me understand the need for the bool?

kajarenc replied:
Sure!

As PM rightly noted in the spec, potentially, email could not be included in the Identity provider response. This is very unlikely to happen because we always by default request `email` scope, and `email` is part of OIDC standard claims https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

However, to protect ourselves from identity providers that don't include email, I decided not to rely on the `email` field to verify successful authentication. Instead, in case of a successful authentication response, we add the `_streamlit_logged_in` field into the `user_info` of `AppSession` (we also set that field to True in case of not empty/placeholder email from the Cloud header). 

Then we compute the `is_authenticated` property method of `experimental_user` based on that explicit value.

To not raise any typing errors, we describe explicitly here that `user_info` could also carry a `bool` value. 
These changes are for mypy, on the runtime time they don't affect what exact values we store in `user_info`.

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
* sort dependencies list, and include upper bound for Authlib
* Add explanatory comment about _streamlit_logged_in form user_info

sfc-gh-kmcgrady reviewed (approved):
Approving with conversation. Just some nits left. Great work!

kajarenc committed:
changes after review

kajarenc committed:
changes after review, add secure attribute to the cookie

kajarenc committed:
Add explanatory comment

kajarenc committed:
Add .clear() method to st.experimental_user

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
ADD default  provider behaviour

kajarenc committed:
fix error texts in tests

kajarenc committed:
remove error handling

kajarenc committed:
fix small issue for client_kwargs for default provider

kajarenc committed:
*Include _streamlit_logged_in field to st.user
* Make is_authenticated a method instead of property

kajarenc committed:
Rename AuthError to StreamlitAuthError

kajarenc committed:
custom KeyError / AttributeError for st.experimental_user

kajarenc committed:
remove "_streamlit_logged_in" from user info dictionary in case if info from Cloud

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Auth authlib ALTERNATIVE API (#9973)

## Describe your changes
API change for Auth feature
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

---------

Co-authored-by: Johannes Rieke <johannes.rieke@gmail.com>
Co-authored-by: Johannes Rieke <johannes.rieke@snowflake.com>

kajarenc committed:
Merge branch 'develop' into auth-authlib-experiments

kajarenc committed:
Adjust e2e tests script with new API

kajarenc committed:
Better exception text

kajarenc committed:
final exception text update

kajarenc committed:
remove high boundary check for Authlib

kajarenc committed:
remove unused comment in proto file

kajarenc committed:
Make None the default value for provider in st.login

kajarenc committed:
small fix in exception formatting
